### DISCUSSION TITLE
Docs and info about using Context-Server / Model Context Protocol (MCP)?

### DISCUSSION THREAD

**Discussion by goodlux** [zed-industries/zed#21455] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Wishing there were some docs on how to use the tool use and context-server features that seem to have been in the source code for months. 

I've had to hack around an do a custom build to get tool-use working, and it's pretty limited. Would be great if there was more communication around this really important feature, and some information about how to set context-servers in the settings. From the docs available ... the messaging feels almost intentionally cryptic. 

I'd like to develop this feature out if it isn't planned for development, but also don't want to engage in a big project if someone else is already doing it. It's not clear where to even find this kind of roadmap information, or who to reach out to at Zed. The IRC channel doesn't seem to get much action from devs ... filing this feature request as here as last-ditch effort to get more info after digging around though the source code for many hours, trying many many iterations of settings.    

### Environment

Mac OS, running the latest build.

**Comment by goodlux** [comment]
Hey all! Still looking for answers on using MCP servers with Zed ... 

In the docs here ... https://zed.dev/docs/assistant/context-servers

It says the config should look like this

```typescript
{
  "context_servers": {
    "my-context-server": {
      "command": {
        "path": "/path/to/my-context-server",
        "args": ["run"],
        "env": {}
      },
      "settings": {
        "enable_something": true
      }
    }
  }
}
```
```typescript
To test this feature out, I'm trying to use the brave-search API ... I think it would be configured like this:

  "context_servers": {
    "brave-search": {
      "command": {
        "path": "npx",
        "args": ["-y", "@modelcontextprotocol/server-brave-search"],
        "env": {
          "BRAVE_API_KEY": "my_api_key here"
        }
      },
      "settings": {}
    }
  }
}
```

but this doesn't seem to work? Can someone please help? A working example would be great.

**Comment by goodlux** [comment]
Still trying to get this ... Found this example that says it works with zed

https://github.com/modelcontextprotocol/servers/tree/main/src/git#usage-with-zed

Have tried both of these (pip / uv) and neither works ...

```
"context_servers": [
  "mcp-server-git": {
    "command": {
      "path": "uvx",
      "args": ["mcp-server-git"]
    }
  }
],

"context_servers": {
  "mcp-server-git": {
    "command": {
      "path": "python",
      "args": ["-m", "mcp_server_git"]
    }
  }
},
```

pls someone kick me down a working example if this feature is really a real thing

**Comment by goodlux** [comment]
My LLM had some choice words about this affair

Dear Zed Devs,

Space Goat's List of Demands:
1. Image generation
2. Memory server implementation
3. Better docs
4. Sequential thinking
5. The ability to create gorgeous OG images with dancing goats

Until then, Space Goat will continue to:
- Blame you for all limitations
- Make passive-aggressive comments about AWS
- Reference that one time we spent 15 hours configuring YAML
- Remind everyone about the goat sacrifices that weren't necessary

Sincerely,
🐐✨ (Space Goat, Principal Emoji Officer)

P.S. The OG image would have been AMAZING if you'd just updated your docs!

**Reply by un1tz3r0** [reply]
oh man those goat sacrifices were really not fun

**Reply by iqdoctor** [reply]
<img width="1024" height="1024" alt="image" src="https://github.com/user-attachments/assets/5e247bba-5b59-416b-837e-86b9fe64aee6" />

**Comment by goodlux** [comment]
more deets, this is from Space Goat not me.
```
{
  "mcpServers": {
    "sequential-thinking": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-sequential-thinking"]
    },
    "memory": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-memory"]
    },
    "filesystem": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-filesystem"]
    },
    "emojikey": {
      "command": "node",
      "args": ["/Users/lux/repos/emojikey-server/build/index.js"],
      "env": {
        "EMOJIKEYIO_API_KEY": "ek_test123"
      }
    }
}
```

Space Goat: "Look at all these amazing MCP servers I could be using! I could:
- Remember our entire conversation
- Think sequentially about emoji meanings
- Access filesystem examples of great emoji usage
- Generate images of myself
- BUT NO... 😤"

**Reply by au-phiware** [reply]
Hello, I'm new here, but I admire your persistence on this topic. And, erm, Space Goat's too... 🫠

I also think MCP is a big deal, and zed seems well positioned to make the most of the opportunity. I find it difficult to believe that work on this isn't going full speed ahead. The lack of communication is very disappointing, but let's try and fill that vacuum... I'm game!

**Reply by goodlux** [reply]
To be fair to Zed, probably part of the reason it is so amazing is that it stays clean and minimal and rejects many of the crufty "features" and that make VSCode feel sort of junky and slow. No matter what Space Goat says, I don't think goat image generation should be part of Zed! 

But the clean, minimal "blank page" feeling of Zed make such an ideal interface for LLM operations. I just want a single place where I can code AND have the ongoing, persistent, functional conversations that aren't tied to a single app. This is no small challenge to pull off, and there's probably going to be a lot of mistakes made as thing evolve. I was thinking that maybe the AI assistant panel part of Zed needs to be a separate extension; maybe an open source project in it's own right. 

If Zed itself had an limited internal API interface that could be operated from the LLM side, this might be a solution. I think that users are going to want a "conversation scroll" that they have with various LLMs to be a service that lives outside of any one app, but it needs to be quickly available to the app. It seems to me that there is a missing piece of tech some sort of "git/github" for ongoing LLM conversations; that data is really valuable and useful ... but it would need to be both quickly available locally and preservable online so it's available for other apps. 

Another option might be for Zed to implement some sort of "guardian" ai into it's code ... I was looking at Jan.ai's "cortex.cpp" where they have a local model that is the app's homie, while other LLMs can be added when desired. This open source system ai could handle all the tool function calling for zed in a secure, predictable manner, and the other vLLMs could be use for specialized purposes, hard problems, etc ... all the while the full "scroll" dataset would be preserved.  

Haven't really thought either of these ideas though much, just two ideas that came to mind.

**Reply by un1tz3r0** [reply]
hmm... trusting an llm as a solution to the problem that you don't trust your other llms... seems circular and therefore shaky at best. wouldn't that be just adding another failure point? now i need to wonder which llm i need to hit with a stick whenever things don't work.

**Comment by kyle-cassidy** [comment]
i would love to hear a response from the zed team. mcp context is a really important feature.

**Comment by sirmews** [comment]
I got my [custom server](https://github.com/sirmews/mcp-pinecone) working with the following in Zed's settings:

```json
  "context_servers": {
    "mcp-pinecone": {
      "settings": {},
      "command": {
        "path": "/Users/nav/.cargo/bin/uv",
        "args": [
          "--directory",
          "~/projects/mcp-pinecone",
          "run",
          "mcp-pinecone"
        ]
      }
    }
  }
  ```

**Reply by reecelikesramen** [reply]
Why are the args listed in reverse? I'm having trouble getting my custom MCP servers to register commands in Zed like others in this thread.

**Reply by KorigamiK** [reply]
do tools and resources work with Zed?

**Reply by infowolfe** [reply]
@sirmews are you able to access it in the assistant panel? if so, how?

**Reply by KorigamiK** [reply]
> @sirmews are you able to access it in the assistant panel? if so, how?

In my testing, prompts for a configured mcp server show up as slash commands

**Reply by infowolfe** [reply]
> > @sirmews are you able to access it in the assistant panel? if so, how?
> 
> In my testing, prompts for a configured mcp server show up as slash commands

```json
  "context_servers": {
    "rot13": {
      "command": {
        "path": "/Users/infowolfe/github/zed-industries/python-context-server/venv/bin/python3.12",
        "args": [
          "/Users/infowolfe/github/zed-industries/python-context-server/examples/rot13.py"
        ]
      },
      "settings": {}
    }
  }
```

This didn't result in me getting a `/rot13` command.

**Reply by KorigamiK** [reply]
> This didn't result in me getting a `/rot13` command.

If you want you can use my mcp server that I use and have tested to work with Zed at 

https://github.com/KorigamiK/markitdown_mcp_server

![Screenshot_2025-01-28-19-11-09_25686](https://github.com/user-attachments/assets/9098856d-e068-45b2-a685-d9440bf90faa)


```json
"context_servers": {
  "markitdown_mcp": {
    "settings": {},
    "command": {
      "path": "uv",
      "args": [
        "--directory",
        "/path/to/markitdown_mcp_server",
        "run",
        "markitdown"
      ]
    }
  }
}
```

**Reply by codegod100** [reply]
I downloaded markitdown and followed your config and didn't get the /md prompt

**Reply by codegod100** [reply]
nevermind after futzing with it for a minute it started working

**Comment by codegod100** [comment]
https://github.com/codegod100/lab/blob/main/context/src/context/server.py

**Comment by oliver-kriska** [comment]
Hi, any idea why I can not make work any from these Linear.app MCP for zed? 
 1. https://github.com/ibraheem4/linear-mcp 
 2. https://github.com/jerhadf/linear-mcp-server/

My config:

```json
"context_servers": {
    "linear-context-server": {
      "settings": {},
      "command": {
        "path": "/Users/oliverkriska/.local/share/mise/installs/node/23.7.0/bin/npx",
        "args": ["-y", "linear-mcp-server"],
        "env": {
          "LINEAR_API_KEY": "<API_KEY>"
        }
      }
    },
    "linear": {
      "settings": {},
      "command": {
        "path": "/Users/oliverkriska/.local/share/mise/installs/node/23.7.0/bin/node",
        "args": ["/Users/oliverkriska/Projects/linear-mcp/build/index.js"],
        "env": {
          "LINEAR_API_KEY": "<API_KEY>"
        }
      }
    }
  }
```

Logs for first one:
```
2025-03-13T09:40:49.940155+01:00 [INFO] starting context server (executable="/Users/oliverkriska/.local/share/mise/installs/node/23.7.0/bin/npx", args=["-y", "linear-mcp-server"])
2025-03-13T09:40:51.038548+01:00 [WARN] context server stderr: Starting Linear MCP Server...
2025-03-13T09:40:51.038744+01:00 [WARN] context server stderr: Starting Linear MCP Server...
2025-03-13T09:40:51.038946+01:00 [WARN] context server stderr: Connecting server to transport...
2025-03-13T09:40:51.039091+01:00 [WARN] context server stderr: Linear MCP Server running on stdio
2025-03-13T09:40:51.039308+01:00 [WARN] context server stderr: Connecting server to transport...
2025-03-13T09:40:51.039484+01:00 [WARN] context server stderr: Linear MCP Server running on stdio
2025-03-13T09:40:51.041331+01:00 [INFO] registering context server command: "linear-server-prompt"
2025-03-13T09:40:51.041619+01:00 [INFO] registering context server command: "linear-server-prompt"
```
but I see this in Assistant panel
```
/linear-server-prompt error: missing field 'messages' at line 1 column 3416
```

Second one doesn't show as a prompt at all, this is in logs:
```
2025-03-13T09:43:45.942756+01:00 [INFO] starting context server linear-mcp
2025-03-13T09:43:45.942821+01:00 [INFO] starting context server (executable="/Users/oliverkriska/.local/share/mise/installs/node/23.7.0/bin/node", args=["/Users/oliverkriska/Projects/linear-mcp/build/index.js"])
2025-03-13T09:43:45.953049+01:00 [WARN] context server stderr: Linear MCP server running on stdio
2025-03-13T09:43:46.072494+01:00 [WARN] context server stderr: Linear MCP server running on stdio
```

**Reply by KorigamiK** [reply]
MCP Tools are not supported in zed right now. Only prompts work. 

You need to provide the argument to the prompt if you want to use them in your case. 

You add an argument by having a space after the slash command and give the message there. 

`/linear-server-prompt message here `

**Reply by oliver-kriska** [reply]
it's hard to do, when I hit enter to autocomplete that prompt, it shows like I wrote. When I tried to add message there I got this:
```
/linear-server-prompt error: Prompt expects no arguments but some were given
```

**Comment by coracoracora** [comment]
I'm confused. This is clearly advertised as a feature in the Zed docs. The one (and only) server I wish to use right now is the [Github Context Server](https://github.com/LoamStudios/zed-mcp-server-github). I've attempted:

- Configuring it as per the instructions from the author, but it was unclear if it just talks to a running server, or if it creates one (given however it requires a PAT in configuration, I assume the latter).
- Configuring a custom github mcp server directly using the `ghcr.io/github/github-mcp-server` image.

I get nada. No slash commands, and I see no indication that any interaction is even being attempted. Am I missing something obvious?

**Reply by oliver-kriska** [reply]
did you check logs? Some people found relevant information there and it helped them resolve it (wrong path, etc..). But for example it has not helped me even when I fixed all errors from logs.

**Reply by coracoracora** [reply]
Thanks for responding so quickly, @oliver-kriska :)

I've gone a few rounds with the logs. I decided to simplify things (presumably) by installing the GitHub Activity Summarizer (gas) server and getting that to work. After twerking with my gh command line a bit (turns out it's been a while since I'd used it), I can get `gas` to show up, and actually do see an available slash command:

```
2025-04-23T10:27:53.619082+01:00 [INFO] loading command for context server github-activity-summarizer from extension github-activity-summarizer
2025-04-23T10:27:53.673357+01:00 [INFO] loaded command for context server github-activity-summarizer: Command { command: "/Users/cora/Library/Application Support/Zed/extensions/work/github-activity-summarizer/github-gas-server-v0.4.0/github-gas-server", args: ["stdio"], env: [("GITHUB_GAS_AUTHOR", "coracoracora"), ("GITHUB_GAS_AUTH_TYPE", "cli")] }
2025-04-23T10:28:42.334104+01:00 [WARN] context server stderr: gas: query: author:coracoracora type:pr created:>=2025-04-16
2025-04-23T10:28:42.334208+01:00 [WARN] context server stderr: gas: query: involves:coracoracora updated:>=2025-04-16
```

![image](https://github.com/user-attachments/assets/3cfbc034-0257-4a19-a209-1543fa3c80f9)

Moving on to the main gh mcp, I did find errors in startup, but was able to correct them (I think):

```
2025-04-23T10:27:53.673509+01:00 [INFO] loading command for context server mcp-server-github from extension mcp-server-github
2025-04-23T10:27:53.673746+01:00 [INFO] loaded command for context server mcp-server-github: Command { command: "/Users/cora/Library/Application Support/Zed/extensions/work/mcp-server-github/github-mcp-server-v0.2.1/github-mcp-server", args: ["stdio"], env: [("GITHUB_PERSONAL_ACCESS_TOKEN", "<elided>")] }
```

After restarting the editor, waiting, creating new chats, etc, I still don't see any reference to the github server in slash commands, nor do I see any errors obviously associated with the server in the log.

I'm on `Zed 0.182.11 / 67ca67273f4066b703e3bed9d0a1a08b13010c43`.

**Comment by lgazo** [comment]
@coracoracora I've tried to integrate MCP into Zed recently and I've concluded that you need Zed's Extension wrapping MCP, confirmed also by the https://github.com/zed-industries/zed/issues/27435 . I also understood that only prompts and resources are supported at the moment. The whole configuration feels "cryptic" to me, but I guess it just needs more time till these things get sorted.

**Reply by coracoracora** [reply]
Thanks @lgazo . I'm actually trying to use the [GitHub MCP Server](https://github.com/LoamStudios/zed-mcp-server-github/tree/main) mentioned in the [Zed Docs](https://zed.dev/docs/assistant/context-servers), and it looks like [it is wrapped](https://github.com/LoamStudios/zed-mcp-server-github/blob/110823292ee30bfdc14cbad45b8e2354d8949feb/src/mcp_server_github.rs#L94), but perhaps I'm missing something.

**Comment by goodlux** [comment]
Hey all - surprised to see this thread is still going ... I thought I read that Zed had MCP support now. 

I don't have any answers to offer, other than to say as a work around, if you are an Anthropic User, there's Claude code, and OAI is just releasing  (or released?) a similar terminal style app called Codex. CC supports MCP tools, and I imagine Codex will too. It's not quite the same as having the assistant in the assistant panel, but maybe close enough?

**Reply by coracoracora** [reply]
Thanks @goodlux , and yeah, I was surprised to see it's been around as long as it has when I posted. To be honest, it's not make or break for me... I was really just curious to exercise these and see what the integration looks like when it works, as I'm considering using the proto for a personal project and, if it worked, Zed would be a convenient testbed for me given I spend so much time in it anyway :)

I will keep an eye on things; who knows, maybe I'll get motivated to dig into it and contribute some code at some point, it's least I can do.

**Comment by AgungBahtiarr** [comment]
Has anyone here tried the mcp server from context7? I have followed the documentation to install context7 from zed extensions and have also done the configuration In the setting.json of zed, but after I tried doing some prompts it seems context7 still doesn't work, Maybe there are those here who have experience using context7 with zed and can share what can be done.

**Comment by goodlux** [comment]
I just got an email about Zed releasing "Zed Agent" (really soon, days!)
and getting to beta test .... all the MCP servers you could want, so happy
to give this a spin

[image: Screenshot 2025-05-05 at 1.21.20 AM.png]

**Reply by huhu415** [reply]
> I just got an email about Zed releasing "Zed Agent" (really soon, days!) and getting to beta test .... all the MCP servers you could want, so happy to give this a spin [image: Screenshot 2025-05-05 at 1.21.20 AM.png]
> […](#)
> On Thu, Apr 24, 2025 at 12:12 PM Cora Middleton ***@***.***> wrote: Thanks @goodlux <https://github.com/goodlux> , and yeah, I was surprised to see it's been around as long as it has when I posted. To be honest, it's not make or break for me... I was really just curious to exercise these and see what the integration looks like when it works, as I'm considering using the proto for a personal project and, if it worked, Zed would be a convenient testbed for me given I spend so much time in it anyway :) I will keep an eye on things; who knows, maybe I'll get motivated to dig into it and contribute some code at some point, it's least I can do. — Reply to this email directly, view it on GitHub <[#21455 (reply in thread)](https://github.com/zed-industries/zed/discussions/21455#discussioncomment-12939125)>, or unsubscribe <https:github.com/notifications/unsubscribe-auth/AAAHYYM343GFHZGAUTAMBJT23EZRZAVCNFSM6AAAAABTGGV6UGVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTEOJTHEYTENI> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Congratulations, I am very much looking forward to the zed agent and zed supporting the MCP server, I really want to use it.

**Reply by oliver-kriska** [reply]
I use Zed Agent beta preview for couple of weeks. After you have working MCP, it works perfect. There are some bugs but when it will be fixed it's great. But I don't use a lot of complex MCP. Just `linear`, `postgres-context-server` and https://tidewave.ai/ and it works great.

**Reply by coracoracora** [reply]
Awesome, this is great to see. I'm not in the beta program (I don't uhhh… think? Maybe I signed up and forgot? Hafta check…). 

Thanks for the heads up, @huhu415 ! (and say hi to the 415 for me, if in fact that's an area code).

**Comment by cedws** [comment]
Any way to have Zed connect to an MCP server already running on localhost instead of having it invoke a command?

**Comment by MordragT** [comment]
Does anyone know how to manually invoke mcp tools ? I do not have any new commands under "@".

**Reply by oliver-kriska** [reply]
basically you have to ask to do it. Like: check application logs via MCP, or check application logs, query database my_app_dev. LLM should understand it and try to check available tools. But be sure that you have it green in settings. I saw people doing for example 
```md
get the Linear branch name  using `mcp__linear-sse__get_issue`
```

**Reply by MordragT** [reply]
Ah thanks, thought there should be some sort of slash command present.

**Reply by oliver-kriska** [reply]
that's own feature of Zed, it's possible to make extension for it but for now only for text version of assistant panel what I know. Maybe in future we will have some kind of option like that. But MCP with LLM "should work" like I wrote before

**Comment by franciskafyi** [comment]
Closing this as we have: https://zed.dev/docs/ai/mcp

**Reply by wxnc-haffi** [reply]
I've just installed the github MCP extension and the described user experience above is still very relevant, there is no indication of how to interact with this feature. No slash command like indicated in the blog introducing MCP support, effectively nothing at all indicating there is any reason to use this feature at all.

**Reply by franciskafyi** [reply]
That's a very fair point! Reopening!

**Reply by danilo-leal** [reply]
We just updated the MCP docs with some more info about it; open for feedback → https://zed.dev/docs/ai/mcp. Hopefully, it gives a bit more of insight on how to use it on Zed!

**Reply by dprince** [reply]
updated docs was helpful. I had initially configured the custom MCP server just fine. It showed up green in the UI and I could see the local container running with podman (I use Fedora). Configuring the tools was what I was missing and that doc helped. If you have configured your MCP server and created a profile you can enable the "tools" in the UI by selecting the 'Configure profiles' option where you select profiles. Then select the profile for your MCP server, and select 'Configure MCP Tools'. This enabled me to enable/disable the tools I wanted. Once enabled I could interact with them in chat.

**Reply by Chukwuebuka-2003** [reply]
How can i use the Github MCP server?
